---
title: "chipse-analysis"
output: html_document
date: '2022-07-04'
---

```{r}

# install the package "ChIPpeakAnno" from Bioconductor:
source("http://bioconductor.org/biocLite.R")
biocLite("ChIPpeakAnno")

```


The function read.table reads an input file and transformed it into a data.frame, the basic data structure in R. Lines starting with # are ignored by default and the option header=TRUE indicates that the first line contains the column names.


```{r}
H3K27ac_Old.df <- read.table("C:/Users/Pushpa Koirala/Documents/datascience/GSM3752856_O-6A-H3K27ac.bed", header=TRUE)
head(healthy.df)


```





2.2 Convert peaks into a GRanges object
A GRanges object form the GenomicRanges packages is a container for genomic locations and their associated annotations. We will use this data structure for the ChIP-seq peak coordinates.


```{r}
require(GenomicRanges) 
H3K27ac_Old <- GRanges(
  H3K27ac_Old.df$chr,
  IRanges(H3K27ac_Old.df$X15045,H3K27ac_Old.df$X15275),
  strand="*"
)
#Adding more data to each peak subsequently
names(H3K27ac_Old) <- paste("gene_peak", 1:nrow(H3K27ac_Old.df), sep="_")
score(H3K27ac_Old) <- H3K27ac_Old.df$X0.017620  
H3K27ac_Old


```


Number of Peaks
```{r}
length(H3K27ac_Old)
```




```{r}

H3K27ac_Old.size <- width(H3K27ac_Old)
summary(H3K27ac_Old.size)

```

A GRanges object can be indexed using the [] operator similar than vectors.

```{r}

H3K27ac_Old <- H3K27ac_Old[width(H3K27ac_Old) <= 2000]

```




```{r}

hist(width(H3K27ac_Old), xlab="H3K27ac_Old peak size", col="gray")

```

What is the distribution of gene ChIP-seq peak p-values?

```{r}
mlPvals <- score(H3K27ac_Old)
hist(mlPvals, xlab="p-value", col="gray")
```

Comparing sample with Alzeimer disease and the old person sample with no Alzeimer disease
Loading the Alzeimer disease Sample for comparision.
```{r}
require(rtracklayer) 
H3K27ac_AD <- import.bed("C:/Users/Pushpa Koirala/Documents/datascience/GSM3752873_AD-6T-H3K27ac.bed")
H3K27ac_AD

```

Fixing the name and score coloumns


```{r}

bp <- barplot(c(length(H3K27ac_AD), length(H3K27ac_Old)), names=c("H3K27ac_AD", "H3K27ac_Old"))
text(bp, c(length(H3K27ac_AD), length(H3K27ac_Old)), labels=c(length(H3K27ac_AD), length(H3K27ac_Old)), pos=1)

```

How many H3K27ac_AD peaks overlap gene peaks?
```{r}
# compute overlaps
ovlHits <- findOverlaps(H3K27ac_AD, H3K27ac_Old)

# show the resulting Hits object
ovlHits
```



```{r}

# get subsets of binding sites
ovl <- subsetByOverlaps(H3K27ac_AD, H3K27ac_Old)

# number of overlaps
length(ovl)

```




```{r}
# as percent
length(ovl) / length(H3K27ac_AD) * 100
```


 A Venn-diagram showing the overlap of of AD and gpeaks. Taking the subset of peaks that are unique to AD and gene using the function setdiff. Then we use the package Vennerable to build a Venn object with the number of peaks in each subset. This object can than be passed to the ‘plot’ function.

```{r}

# get subsets of binding sites
H3K27ac_AD.uniq <- setdiff(H3K27ac_AD, H3K27ac_Old)
H3K27ac_Old.uniq <- setdiff(H3K27ac_Old, H3K27ac_AD)

```

```{r}
install.packages("Vennerable", repos="http://R-Forge.R-project.org")
```

```{r}

# plot overlap of binding sites as Venn diagram
require(Vennerable)
library(reshape)
library(Vennerable)

# build objects with the numbers of sites in the subsets
venn <- Venn(SetNames=c("H3K27ac_AD", "H3K27ac_Old"), 
    Weight=c(
        '10'=length(H3K27ac_AD.uniq), 
        '11'=length(ovl), 
        '01'=length(H3K27ac_Old.uniq)
    )
)

# plot Venn Diagram
plot(venn)

```






Functional Annotation


```{r}
# load gene annotation from UCSC for human genome build hg18
require(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene 
```


We will use the ChIPpeakAnno package. It has the function assignChromosomeRegion for calculating these overlaps by using a annotation database.
```{r}
require(ChIPpeakAnno) 

library(ChIPpeakAnno)
# laod package to annotate peaks with genes
# calucluate the overlap with features
H3K27ac_AD.features <- assignChromosomeRegion(H3K27ac_AD, TxDb=txdb, nucleotideLevel=FALSE)
# show the results
H3K27ac_AD.features
```


Alzeihmer Samples peaks that overlap the promotor region. First calculating the promotors as GRanges object

```{r}
# get all genes from the data base as GRanges object
genes <- genes(txdb)
# take the the region arround the gene start as promoter
prom <- promoters(genes, upstream=2000, downstream=200)
prom

```


Now Calculating the subset of AD peaks that overlap a promotor region
```{r}
# get only those ER peaks that overlap a promoter region
H3K27ac_ADatProm <- subsetByOverlaps(H3K27ac_AD, prom)
# subset size
length(H3K27ac_ADatProm)
```


```{r}
# percent of all ER peaks
length(H3K27ac_ADatProm) / length(AD) * 100


```


Subset of genes that have a peak in their promotor regions

```{r}
# We search for overlap between ER peaks and promoters
H3K27ac_ADatProm.Hits = findOverlaps(H3K27ac_AD, prom)

H3K27ac_ADprom = genes[subjectHits(H3K27ac_ADatProm.Hits)]

# take only unique ids
gene1.ids <- unique(names(H3K27ac_ADprom))

# write names to an output file
write.table(gene1.ids, file="H3K27ac_AD_regulated_genes.txt", quote=FALSE, row.names=FALSE, col.names=FALSE)

```


```{r}



```

