---
title: "main_notebook"
output: html_document
date: '2022-06-20'
---

# Data Science \| Final Project \| Group 02

# Identify epigenetic alterations associated with Alzheimer's disease and classification of gene expressions between healthy and sick patients

**Christina Kirschbaum, Pushpa Koirala, Melika Moradi**

## Our Project

Alzheimer's disease is the most prevalent kind of dementia and a fatal brain ailment. More study into this illness might lead to a better understanding of the condition and more effective treatment options.

In this project, ChIP-seq data for H3K27ac, H3K9ac, H3K122ac and H3K4me1 as well as RNA-seq data from the from the lateral temporal lobe of the human brain for young healthy patients, old heathy patients and patients with Alzheimer's disease will be analyzed and the differences between normal aging and cognitive impairment will be explored. The epigenomic or transcriptomic profiles will be analyzed to find relevant epigenetic alterations associated with the disease and help to better understand the molecular pathophysiology underlying.

Afterwards, with Machine Learning models the presence and absence of Alzheimer's disease based on the data. The models will be built with Support Vector Machines and Random Forest. Additionally, clustering methods will be tested.

Finally, we will look at the functional enrichment of the ChIP-seq data and will add a STRING analysis for the RNA-seq data.

For our project we were inspirated by the paper: Nativio R, Lan Y, Donahue G et al. ["An integrated multi-omics approach identifies epigenetic alterations associated with Alzheimer's disease."](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8098004/)

The data is derived from GEO, [Series GSE153875](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE153875), for the CHiP-seq data and the raw RNA-seq data [Series GSE159699](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA670209&o=acc_s%3Aa).

For more detailed information about the data, please refer to the README of the repository.

## Data Integration RNA-seq

This part can be found in the Python notebook [notebook.ipynb](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/notebook.ipynb) in the repository, in the section ***Data Integration RNA-seq***.

## Data Integration ChIP-seq

This part can be found in the Python notebook [notebook.ipynb](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/notebook.ipynb) in the repository, in the section ***Data Integration ChIP-seq***.

## Preprocessing RNA-seq

This part can be found in the Python notebook [notebook.ipynb](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/notebook.ipynb) in the repository, in the section ***Preprocessing RNA-seq***.

## Quality Control ChIP-seq

This part can be found in the Python notebook [notebook.ipynb](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/notebook.ipynb) in the repository, in the section ***Quality Control ChIP-seq***.

## Data Analysis ChIP-seq

The main part of the ChIP-seq analysis can be found in the following section. For the deeptools part, refer the Python notebook [notebook.ipynb](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/notebook.ipynb) in the repository, in the section ***Data Analysis ChIP-seq***.

For the example code for `computeMatrix` and `plotHeatmap` of deeptools with H3K27ac, take  a look at [commands.txt](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/commands.txt).

### H3K27ac ChIPpeakAnno

```{r}
library(BiocManager)
install("ChIPQC")
require(rtracklayer)
```


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()
```


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ChIPpeakAnno")

```

The function read.table reads an input file and transformed it into a data.frame, the basic data structure in R. Lines starting with \# are ignored by default and the option header=TRUE indicates that the first line contains the column names.

```{r}
#The function read.table reads an input file and transformed it into a data.frame, the basic data structure in R. 
#Lines starting with # are ignored by default and the option header=TRUE indicates that the first line contains the column names.
H3K27ac_Old.df <- read.table("C:/Users/Pushpa Koirala/Documents/datascience/GSM3752856_O-6A-H3K27ac.bed", header=TRUE)
head(H3K27ac_Old.df)
```

```{r}
#Convert peaks into a GRanges object
#A GRanges object form the GenomicRanges packages is a container for genomic locations and their associated annotations. 
#We will use this data structure for the ChIP-seq peak coordinates
require(GenomicRanges) 
H3K27ac_Old <- GRanges(
  H3K27ac_Old.df$chr,
  IRanges(H3K27ac_Old.df$X15045,H3K27ac_Old.df$X15275),
  strand="*"
)
#Adding more data to each peak subsequently
names(H3K27ac_Old) <- paste("gene_peak", 1:nrow(H3K27ac_Old.df), sep="_")
score(H3K27ac_Old) <- H3K27ac_Old.df$X0.017620  
H3K27ac_Old
```

```{r}
#Number of Peaks
length(H3K27ac_Old)
```

```{r}
H3K27ac_Old.size <- width(H3K27ac_Old)
summary(H3K27ac_Old.size)
```

```{r}
#A GRanges object can be indexed using the [] operator similar than vectors.
H3K27ac_Old <- H3K27ac_Old[width(H3K27ac_Old) <= 2000]
```

```{r}
hist(width(H3K27ac_Old), xlab="H3K27ac_Old peak size", col="gray")
```

```{r}
#What is the distribution of gene ChIP-seq peak p-values?
mlPvals <- score(H3K27ac_Old)
hist(mlPvals, xlab="p-value", col="gray")
```

```{r}
#Comparing sample with Alzeimer disease and the old person sample with no Alzeimer disease and loading the Alzeimer disease Sample for comparision.
require(rtracklayer) 
H3K27ac_AD <- import.bed("C:/Users/Pushpa Koirala/Documents/datascience/GSM3752873_AD-6T-H3K27ac.bed")
H3K27ac_AD
```

```{r}
#Fixing the name and score coloumns
bp <- barplot(c(length(H3K27ac_AD), length(H3K27ac_Old)), names=c("H3K27ac_AD", "H3K27ac_Old"))
text(bp, c(length(H3K27ac_AD), length(H3K27ac_Old)), labels=c(length(H3K27ac_AD), length(H3K27ac_Old)), pos=1)
```

```{r}
#How many H3K27ac_AD peaks overlap gene peaks?
# compute overlaps
ovlHits <- findOverlaps(H3K27ac_AD, H3K27ac_Old)

# show the resulting Hits object
ovlHits
```

```{r}
# get subsets of binding sites
ovl <- subsetByOverlaps(H3K27ac_AD, H3K27ac_Old)

# number of overlaps
length(ovl)
```

```{r}
# as percent
length(ovl) / length(H3K27ac_AD) * 100
```

```{r}
# get subsets of binding sites. A Venn-diagram showing the overlap of of AD and gpeaks. Taking the subset of peaks that are unique to AD and gene using the function setdiff. Then we use the package Vennerable to build a Venn object with the number of peaks in each subset. This object can than be passed to the ‘plot’ function.
H3K27ac_AD.uniq <- setdiff(H3K27ac_AD, H3K27ac_Old)
H3K27ac_Old.uniq <- setdiff(H3K27ac_Old, H3K27ac_AD)
```

```{r}
install.packages("Vennerable", repos="http://R-Forge.R-project.org")
```

```{r}
# plot overlap of binding sites as Venn diagram
require(Vennerable)
library(reshape)
library(Vennerable)

# build objects with the numbers of sites in the subsets
venn <- Venn(SetNames=c("H3K27ac_AD", "H3K27ac_Old"), 
    Weight=c(
        '10'=length(H3K27ac_AD.uniq), 
        '11'=length(ovl), 
        '01'=length(H3K27ac_Old.uniq)
    )
)

# plot Venn Diagram
plot(venn)
```

#### H3K27ac ChIPpeakAnno Functional Enrichment

```{r}
# load gene annotation from UCSC for human genome build hg18
require(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene 
```

```{r}
require(ChIPpeakAnno)   # load package to annotate peaks with genes.We will use the ChIPpeakAnno package. It has the function assignChromosomeRegion for calculating these overlaps by using a annotation database.
# calucluate the overlap with features
H3K27ac_AD.features <- assignChromosomeRegion(H3K27ac_AD, TxDb=txdb, nucleotideLevel=FALSE)
# show the results
H3K27ac_AD.features
```

```{r}
# get all genes from the data base as GRanges object.Alzeihmer Sample peaks that overlap the promotor region. First calculating the promotors as GRanges object
genes <- genes(txdb)
# take the the region arround the gene start as promoter
prom <- promoters(genes, upstream=2000, downstream=200)
prom
```

```{r}
# get only those peaks that overlap a promoter region.Now Calculating the subset of AD peaks that overlap a promotor region
H3K27ac_ADatProm <- subsetByOverlaps(H3K27ac_AD, prom)
# subset size
length(H3K27ac_ADatProm)
```

```{r}
# percent of all peaks
length(H3K27ac_ADatProm) / length(AD) * 100
```

```{r}
# We search for overlap between peaks and promoters.
#Subset of genes that have a peak in their promotor regions
H3K27ac_ADatProm.Hits = findOverlaps(H3K27ac_AD, prom)

H3K27ac_ADprom = genes[subjectHits(H3K27ac_ADatProm.Hits)]

# take only unique ids
gene1.ids <- unique(names(H3K27ac_ADprom))
```

### H3K4me1 with ChIPseeker

First, we import all the librarys that are needed to do the analysis with the Biocondutor package [ChIPseeker](https://bioconductor.org/packages/release/bioc/html/ChIPseeker.html). We abbreviate `TxDb.Hsapiens.UCSC.hg19.knownGene` to `txdb` to access it easier.

```{r echo = T, results = 'hide'}
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
library(rafalib)
library(ggplot2)
```

In the next step, we read the BED files for H3K4me1 and split them into AD, old and young.

```{r}
path <- "data/H3K4me1_bed/"

files <- list.files(path = path)

adfile <- grep("AD", files, value = TRUE)
adfile <- grep(".bed", adfile, value = TRUE)
for (i in 1:length(adfile)) assign(adfile[i], readPeakFile(paste(path,adfile[i], sep="")))

oldfile <- grep("O", files, value = TRUE)
oldfile <- grep(".bed", oldfile, value = TRUE)
for (i in 1:length(oldfile)) assign(oldfile[i], readPeakFile(paste(path,oldfile[i], sep="")))

youngfile <- grep("Y", files, value = TRUE)
youngfile <- grep(".bed", youngfile, value = TRUE)
for (i in 1:length(youngfile)) assign(youngfile[i], readPeakFile(paste(path,youngfile[i], sep="")))
```

In the following three chunks, we build GRangesLists for every group.

```{r}
ad <- GRangesList(`AD-1T`=`GSM3752920_AD-1T-H3K4me1.bed`,
                  `AD-1A`=`GSM3752921_AD-1A-H3K4me1.bed`,
                  `AD-2T`=`GSM3752922_AD-2T-H3K4me1.bed`,
                  `AD-2A`=`GSM3752923_AD-2A-H3K4me1.bed`,
                  `AD-3T`=`GSM3752924_AD-3T-H3K4me1.bed`,
                  `AD-5T`=`GSM3752925_AD-5T-H3K4me1.bed`,
                  `AD-3A`=`GSM3752926_AD-3A-H3K4me1.bed`,
                  `AD-5A`=`GSM3752927_AD-5A-H3K4me1.bed`,
                  `AD-8T`=`GSM3752928_AD-8T-H3K4me1.bed`,
                  `AD-6T`=`GSM3752929_AD-6T-H3K4me1.bed`,
                  `AD-7T`=`GSM3752930_AD-7T-H3K4me1.bed`)
```

```{r}
old <- GRangesList(`O-8A`=`GSM3752910_O-8A-H3K4me1.bed`,
                   `O-10T`=`GSM3752911_O-10T-H3K4me1.bed`,
                   `O-6A`=`GSM3752912_O-6A-H3K4me1.bed`,
                   `O-11T`=`GSM3752913_O-11T-H3K4me1.bed`,
                   `O-7A`=`GSM3752914_O-7A-H3K4me1.bed`,
                   `O-13T`=`GSM3752915_O-13T-H3K4me1.bed`,
                   `O-14T`=`GSM3752916_O-14T-H3K4me1.bed`,
                   `O-9A`=`GSM3752917_O-9A-H3K4me1.bed`,
                   `O-10A`=`GSM3752918_O-10A-H3K4me1.bed`,
                   `O-11A`=`GSM3752919_O-11A-H3K4me1.bed`)
```

```{r}
young <- GRangesList(`Y-15T`=`GSM3752902_Y-15T-H3K4me1.bed`,
                     `Y-17T`=`GSM3752903_Y-17T-H3K4me1.bed`,
                     `Y-13A`=`GSM3752904_Y-13A-H3K4me1.bed`,
                     `Y-18T`=`GSM3752905_Y-18T-H3K4me1.bed`,
                     `Y-14A`=`GSM3752906_Y-14A-H3K4me1.bed`,
                     `Y-19T`=`GSM3752907_Y-19T-H3K4me1.bed`,
                     `Y-15A`=`GSM3752908_Y-15A-H3K4me1.bed`,
                     `Y-16A`=`GSM3752909_Y-16A-H3K4me1.bed`)
```

The next four chunks will generate coverage plots for the peak over the chromosomes. The first 3 chunks will compare within the groups, while the fourth chunk will compare over all groups.

`p` will show an overlapping plot for all samples, while adding `facet_grid()` will generate a seperate plot for every sample.

Note that the `weightCol` parameter is necessary to distinguish the coverage easier. Plotting the coverage plot will take some time.

```{r}
p <- covplot(ad, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(old, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(young, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(c(ad, old, young), weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

With the function `lapply` we apply `annotatePeak` to every item in ad, old and young. 

```{r}
adList <- lapply(ad, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
oldList <- lapply(old, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
youngList <- lapply(young, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
```

With the function `plotAnnoBar` we can look at the feature distribution for any conspicous occurences. 

```{r}
plotAnnoBar(adList)
plotAnnoBar(oldList)
plotAnnoBar(youngList)
```

The same can be done for transcription factor binding loci with `plotDistToTSS`.

```{r}
plotDistToTSS(adList)
plotDistToTSS(oldList)
plotDistToTSS(youngList)
```

In the following, we define the promoter and then generate the tagMatrix for every sample. Be aware that this step will take some time.

```{r}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

adMatrix <- lapply(ad, getTagMatrix, windows=promoter)
oldMatrix <- lapply(old, getTagMatrix, windows=promoter)
youngMatrix <- lapply(young, getTagMatrix, windows=promoter)
```

With the tagMatrix, we can plot the average profile for every sample. Here, we generate a plot per group with a separate curve for every sample. 

Adding `facet = 'row'` will show each curve in a separate plot. With small modifications we can plot a single sample (for example `adMatrix[1]`) or a comparison by adding one sample per group (for example `c(adMatrix[1], oldMatrix[1], youngMatrix[1])`).

Be aware that plotting these will take some time.

```{r}
plotAvgProf(adMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(oldMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(youngMatrix, xlim = c(-3000,3000), conf = 0.95)
```

#### H3K4me1 Functional Enrichment

To get an insight in the functional enrichment of the ChIP-seq data, we wanted to take a look at KEGG (Kyoto Encyclopedia of Genes and Genomes) which annotates to pathways.

In this version of the code, it is done for only the `adList` samples. However, `oldList` and `youngList` concluded in similar results, therefore not much can be concluded for the functional enrichment.

```{r}
library(clusterProfiler)

genes = lapply(adList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))

compKEGG <- compareCluster(geneCluster   = genes,
                           fun           = "enrichKEGG",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH")

dotplot(compKEGG, 
        showCategory = 10, 
        title = "KEGG Pathway Enrichment Analysis")
```

Because there was not much information in the KEGG approach, we also took a look at the GO (Gene Ontology) enrichment which annotates biological processes and molecular functions.

Again, there was not much information in the results.

```{r}
library(org.Hs.eg.db)
compGO <- compareCluster(geneCluster   = genes,
                         fun           = "enrichGO",
                         pvalueCutoff  = 0.05,
                         pAdjustMethod = "BH",
                         OrgDb = "org.Hs.eg.db")

dotplot(compGO, 
        showCategory = 10, 
        title = "GO Enrichment Analysis")
```

### H3K9ac with ChIPseeker

In the following, all the steps described for H3K4me1 will be repeated accordingly for H3K9ac.

```{r}
path <- "data/H3K9ac_bed/"

files <- list.files(path = path)

adfile <- grep("AD", files, value = TRUE)
adfile <- grep(".bed", adfile, value = TRUE)
for (i in 1:length(adfile)) assign(adfile[i], readPeakFile(paste(path,adfile[i], sep="")))

oldfile <- grep("O", files, value = TRUE)
oldfile <- grep(".bed", oldfile, value = TRUE)
for (i in 1:length(oldfile)) assign(oldfile[i], readPeakFile(paste(path,oldfile[i], sep="")))

youngfile <- grep("Y", files, value = TRUE)
youngfile <- grep(".bed", youngfile, value = TRUE)
for (i in 1:length(youngfile)) assign(youngfile[i], readPeakFile(paste(path,youngfile[i], sep="")))
```

```{r}
ad <- GRangesList(`AD-1T`=`GSM3752834_AD-1T-H3K9ac.bed`,
                  `AD-1A`=`GSM3752835_AD-1A-H3K9ac.bed`,
                  `AD-2T`=`GSM3752836_AD-2T-H3K9ac.bed`,
                  `AD-2A`=`GSM3752837_AD-2A-H3K9ac.bed`,
                  `AD-3T`=`GSM3752838_AD-3T-H3K9ac.bed`,
                  `AD-5T`=`GSM3752839_AD-5T-H3K9ac.bed`,
                  `AD-3A`=`GSM3752840_AD-3A-H3K9ac.bed`,
                  `AD-5A`=`GSM3752841_AD-5A-H3K9ac.bed`,
                  `AD-8T`=`GSM3752842_AD-8T-H3K9ac.bed`,
                  `AD-6T`=`GSM3752843_AD-6T-H3K9ac.bed`,
                  `AD-7T`=`GSM3752844_AD-7T-H3K9ac.bed`)
```

```{r}
old <- GRangesList(`O-8A`=`GSM3752824_O-8A-H3K9ac.bed`,
                   `O-10T`=`GSM3752825_O-10T-H3K9ac.bed`,
                   `O-6A`=`GSM3752826_O-6A-H3K9ac.bed`,
                   `O-11T`=`GSM3752827_O-11T-H3K9ac.bed`,
                   `O-7A`=`GSM3752828_O-7A-H3K9ac.bed`,
                   `O-13T`=`GSM3752829_O-13T-H3K9ac.bed`,
                   `O-14T`=`GSM3752830_O-14T-H3K9ac.bed`,
                   `O-9A`=`GSM3752831_O-9A-H3K9ac.bed`,
                   `O-10A`=`GSM3752832_O-10A-H3K9ac.bed`,
                   `O-11A`=`GSM3752833_O-11A-H3K9ac.bed`)
```

```{r}
young <- GRangesList(`Y-15T`=`GSM3752815_Y-15T-H3K9ac.bed`,
                     `Y-12A`=`GSM3752816_Y-12A-H3K9ac.bed`,
                     `Y-17T`=`GSM3752817_Y-17T-H3K9ac.bed`,
                     `Y-13A`=`GSM3752818_Y-13A-H3K9ac.bed`,
                     `Y-18T`=`GSM3752819_Y-18T-H3K9ac.bed`,
                     `Y-14A`=`GSM3752820_Y-14A-H3K9ac.bed`,
                     `Y-19T`=`GSM3752821_Y-19T-H3K9ac.bed`,
                     `Y-15A`=`GSM3752822_Y-15A-H3K9ac.bed`,
                     `Y-16A`=`GSM3752823_Y-16A-H3K9ac.bed`)
```

```{r}
p <- covplot(ad, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(old, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(young, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(c(ad, old, young), weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
adList <- lapply(ad, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
oldList <- lapply(old, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
youngList <- lapply(young, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
```

```{r}
plotAnnoBar(adList)
plotAnnoBar(oldList)
plotAnnoBar(youngList)
```

```{r}
plotDistToTSS(adList)
plotDistToTSS(oldList)
plotDistToTSS(youngList)
```

```{r}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

adMatrix <- lapply(ad, getTagMatrix, windows=promoter)
oldMatrix <- lapply(old, getTagMatrix, windows=promoter)
youngMatrix <- lapply(young, getTagMatrix, windows=promoter)
```

```{r}
plotAvgProf(adMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(oldMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(youngMatrix, xlim = c(-3000,3000), conf = 0.95)
```

#### H3K9ac Functional Enrichment

```{r}
genes = lapply(adList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))

compKEGG <- compareCluster(geneCluster   = genes,
                           fun           = "enrichKEGG",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH")

dotplot(compKEGG, 
        showCategory = 10, 
        title = "KEGG Pathway Enrichment Analysis")
```

```{r}
compGO <- compareCluster(geneCluster   = genes,
                           fun           = "enrichGO",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH",
                         OrgDb = "org.Hs.eg.db")

dotplot(compGO, 
        showCategory = 10, 
        title = "GO Enrichment Analysis")
```

### H3K27ac with ChIPseeker

In the following, all the steps described for H3K4me1 will be repeated accordingly for H3K27ac.

```{r}
path <- "data/H3K27ac_bed/"

files <- list.files(path = path)

adfile <- grep("AD", files, value = TRUE)
adfile <- grep(".bed", adfile, value = TRUE)
for (i in 1:length(adfile)) assign(adfile[i], readPeakFile(paste(path,adfile[i], sep="")))

oldfile <- grep("O", files, value = TRUE)
oldfile <- grep(".bed", oldfile, value = TRUE)
for (i in 1:length(oldfile)) assign(oldfile[i], readPeakFile(paste(path,oldfile[i], sep="")))

youngfile <- grep("Y", files, value = TRUE)
youngfile <- grep(".bed", youngfile, value = TRUE)
for (i in 1:length(youngfile)) assign(youngfile[i], readPeakFile(paste(path,youngfile[i], sep="")))
```

```{r}
ad <- GRangesList(`AD-1T`=`GSM3752864_AD-1T-H3K27ac.bed`,
                  `AD-1A`=`GSM3752865_AD-1A-H3K27ac.bed`,
                  `AD-2T`=`GSM3752866_AD-2T-H3K27ac.bed`,
                  `AD-2A`=`GSM3752867_AD-2A-H3K27ac.bed`,
                  `AD-3T`=`GSM3752868_AD-3T-H3K27ac.bed`,
                  `AD-5T`=`GSM3752869_AD-5T-H3K27ac.bed`,
                  `AD-3A`=`GSM3752870_AD-3A-H3K27ac.bed`,
                  `AD-5A`=`GSM3752871_AD-5A-H3K27ac.bed`,
                  `AD-8T`=`GSM3752872_AD-8T-H3K27ac.bed`,
                  `AD-6T`=`GSM3752873_AD-6T-H3K27ac.bed`,
                  `AD-7T`=`GSM3752874_AD-7T-H3K27ac.bed`)
```

```{r}
old <- GRangesList(`O-8A`=`GSM3752854_O-8A-H3K27ac.bed`,
                   `O-10T`=`GSM3752855_O-10T-H3K27ac.bed`,
                   `O-6A`=`GSM3752856_O-6A-H3K27ac.bed`,
                   `O-11T`=`GSM3752857_O-11T-H3K27ac.bed`,
                   `O-7A`=`GSM3752858_O-7A-H3K27ac.bed`,
                   `O-13T`=`GSM3752859_O-13T-H3K27ac.bed`,
                   `O-14T`=`GSM3752860_O-14T-H3K27ac.bed`,
                   `O-9A`=`GSM3752861_O-9A-H3K27ac.bed`,
                   `O-10A`=`GSM3752862_O-10A-H3K27ac.bed`,
                   `O-11A`=`GSM3752863_O-11A-H3K27ac.bed`)
```

```{r}
young <- GRangesList(`Y-15T`=`GSM3752845_Y-15T-H3K27ac.bed`,
                     `Y-12A`=`GSM3752846_Y-12A-H3K27ac.bed`,
                     `Y-17T`=`GSM3752847_Y-17T-H3K27ac.bed`,
                     `Y-13A`=`GSM3752848_Y-13A-H3K27ac.bed`,
                     `Y-18T`=`GSM3752849_Y-18T-H3K27ac.bed`,
                     `Y-14A`=`GSM3752850_Y-14A-H3K27ac.bed`,
                     `Y-19T`=`GSM3752851_Y-19T-H3K27ac.bed`,
                     `Y-15A`=`GSM3752852_Y-15A-H3K27ac.bed`,
                     `Y-16A`=`GSM3752853_Y-16A-H3K27ac.bed`)
```

```{r}
p <- covplot(ad, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(old, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(young, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(c(ad, old, young), weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
adList <- lapply(ad, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
oldList <- lapply(old, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
youngList <- lapply(young, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
```

```{r}
plotAnnoBar(adList)
plotAnnoBar(oldList)
plotAnnoBar(youngList)
```

```{r}
plotDistToTSS(adList)
plotDistToTSS(oldList)
plotDistToTSS(youngList)
```

```{r}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

adMatrix <- lapply(ad, getTagMatrix, windows=promoter)
oldMatrix <- lapply(old, getTagMatrix, windows=promoter)
youngMatrix <- lapply(young, getTagMatrix, windows=promoter)
```

```{r}
plotAvgProf(adMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(oldMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(youngMatrix, xlim = c(-3000,3000), conf = 0.95)
```

#### H3K27ac ChIPseeker Functional Enrichment

```{r}
genes = lapply(adList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))

compKEGG <- compareCluster(geneCluster   = genes,
                           fun           = "enrichKEGG",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH")

dotplot(compKEGG, 
        showCategory = 10, 
        title = "KEGG Pathway Enrichment Analysis")
```

```{r}
compGO <- compareCluster(geneCluster   = genes,
                         fun           = "enrichGO",
                         pvalueCutoff  = 0.05,
                         pAdjustMethod = "BH",
                         OrgDb = "org.Hs.eg.db")

dotplot(compGO, 
        showCategory = 10, 
        title = "GO Enrichment Analysis")
```

### H3K122ac with ChIPseeker

In the following, all the steps described for H3K4me1 will be repeated accordingly for H3K122ac.

```{r}
path <- "data/H3K122ac_bed/"

files <- list.files(path = path)

adfile <- grep("AD", files, value = TRUE)
adfile <- grep(".bed", adfile, value = TRUE)
for (i in 1:length(adfile)) assign(adfile[i], readPeakFile(paste(path,adfile[i], sep="")))

oldfile <- grep("O", files, value = TRUE)
oldfile <- grep(".bed", oldfile, value = TRUE)
for (i in 1:length(oldfile)) assign(oldfile[i], readPeakFile(paste(path,oldfile[i], sep="")))

youngfile <- grep("Y", files, value = TRUE)
youngfile <- grep(".bed", youngfile, value = TRUE)
for (i in 1:length(youngfile)) assign(youngfile[i], readPeakFile(paste(path,youngfile[i], sep="")))
```

```{r}
ad <- GRangesList(`AD-1T`=`GSM3752893_AD-1T-H3K122ac.bed`,
                  `AD-1A`=`GSM3752894_AD-1A-H3K122ac.bed`,
                  `AD-2T`=`GSM3752895_AD-2T-H3K122ac.bed`,
                  `AD-2A`=`GSM3752896_AD-2A-H3K122ac.bed`,
                  `AD-5T`=`GSM3752897_AD-5T-H3K122ac.bed`,
                  `AD-3A`=`GSM3752898_AD-3A-H3K122ac.bed`,
                  `AD-5A`=`GSM3752898_AD-3A-H3K122ac.bed`,
                  `AD-8T`=`GSM3752900_AD-8T-H3K122ac.bed`,
                  `AD-7T`=`GSM3752901_AD-7T-H3K122ac.bed`)
```

```{r}
old <- GRangesList(`O-8A`=`GSM3752883_O-8A-H3K122ac.bed`,
                   `O-10T`=`GSM3752884_O-10T-H3K122ac.bed`,
                   `O-6A`=`GSM3752885_O-6A-H3K122ac.bed`,
                   `O-11T`=`GSM3752886_O-11T-H3K122ac.bed`,
                   `O-7A`=`GSM3752887_O-7A-H3K122ac.bed`,
                   `O-13T`=`GSM3752888_O-13T-H3K122ac.bed`,
                   `O-14T`=`GSM3752889_O-14T-H3K122ac.bed`,
                   `O-9A`=`GSM3752890_O-9A-H3K122ac.bed`,
                   `O-10A`=`GSM3752891_O-10A-H3K122ac.bed`,
                   `O-11A`=`GSM3752892_O-11A-H3K122ac.bed`)
```

```{r}
young <- GRangesList(`Y-15T`=`GSM3752875_Y-15T-H3K122ac.bed`,
                     `Y-17T`=`GSM3752876_Y-17T-H3K122ac.bed`,
                     `Y-13A`=`GSM3752877_Y-13A-H3K122ac.bed`,
                     `Y-18T`=`GSM3752878_Y-18T-H3K122ac.bed`,
                     `Y-14A`=`GSM3752879_Y-14A-H3K122ac.bed`,
                     `Y-19T`=`GSM3752880_Y-19T-H3K122ac.bed`,
                     `Y-15A`=`GSM3752881_Y-15A-H3K122ac.bed`,
                     `Y-16A`=`GSM3752882_Y-16A-H3K122ac.bed`)
```

```{r}
p <- covplot(ad, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(old, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(young, weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
p <- covplot(c(ad, old, young), weightCol = 'V5')
print(p)

p +
  facet_grid(chr ~ .id)
```

```{r}
adList <- lapply(ad, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
oldList <- lapply(old, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
youngList <- lapply(young, annotatePeak, TxDb = txdb,
                 tssRegion = c(-3000,3000), verbose=FALSE)
```

```{r}
plotAnnoBar(adList)
plotAnnoBar(oldList)
plotAnnoBar(youngList)
```

```{r}
plotDistToTSS(adList)
plotDistToTSS(oldList)
plotDistToTSS(youngList)
```

```{r}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

adMatrix <- lapply(ad, getTagMatrix, windows=promoter)
oldMatrix <- lapply(old, getTagMatrix, windows=promoter)
youngMatrix <- lapply(young, getTagMatrix, windows=promoter)
```

```{r}
plotAvgProf(adMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(oldMatrix, xlim = c(-3000,3000), conf = 0.95)
plotAvgProf(youngMatrix, xlim = c(-3000,3000), conf = 0.95)
```

#### H3K122ac Functional Enrichment

```{r}
genes = lapply(adList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))

compKEGG <- compareCluster(geneCluster   = genes,
                           fun           = "enrichKEGG",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH")

dotplot(compKEGG, 
        showCategory = 10, 
        title = "KEGG Pathway Enrichment Analysis")
```

```{r}
compGO <- compareCluster(geneCluster   = genes,
                           fun           = "enrichGO",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH",
                         OrgDb = "org.Hs.eg.db")

dotplot(compGO, 
        showCategory = 10, 
        title = "GO Enrichment Analysis")
```

## Data Analysis RNA-seq

For the RNA-seq Data Analysis, a Differential Gene Expression analysis is done with the Biocondutor package [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html).

In the following chunks, we import the needed librarys for the Differential Gene Expression analysis. The first chunk imports the librarys for the analysis, the second chunk imports the librarys for visualization.

```{r echo = T, results = 'hide'}
library(tidyverse)
library(DESeq2)
```

```{r echo = T, results = 'hide'}
library(rafalib)
library(ggplot2)
library(vsn)
library(pheatmap)
library(RColorBrewer)
```

### STAR

First of all, we import the RNA-seq count matrix. Here, the original count matrix from the Nativio et al. paper where the data was aligned with STAR is loaded.

```{r}
star <- read.table(
  'data/GSE159699_summary_count.star.txt', 
  header = TRUE, 
  sep = '\t',
  row.names = 'refGene'
  )
head(star)
```

In the next step, we load the corresponding metafile. The metafile was written by us.

```{r}
coldata_star <- read.csv(
  'data/star_col.csv',
  row.names = 1,
  sep = ';'
  )

head(coldata_star)
```

Here, we change the columnnames of of the count matrix to equal the rownames of the metafile, because the metafile only includes the sample identifier as name.

```{r}
i <- 1
for (name in colnames(star)){
  colnames(star)[colnames(star) == name] <- rownames(coldata_star)[i]
  i = i + 1
}
```

Now, we check the form of our data. In detail, we check if all rows from the metafile are also available as columns of the count matrix and if the rownames of the metafile are the same as the columnnames of the count matrix. If not both of the following commands return TRUE, DESeq is not able to work with the files.

```{r}
all(rownames(coldata_star) %in% colnames(star))
all(rownames(coldata_star) == colnames(star))
```

In the following, we first factor the groups of the metafile because otherwise there will be a warning and build a DESeqDataSet afterwards.

```{r}
coldata_star$group <- factor(coldata_star$group)

dds_star <- DESeqDataSetFromMatrix(
  countData = star,
  colData = coldata_star,
  design = ~ group
)

dds_star
```

We perform a pre-filtering which removes rows with less than 10 counts from the DESeqDataSet. Here, about 4000 rows are removed. Note that a more strict filtering will be applied when we use `results()` later on.

```{r}
keep <- rowSums(counts(dds_star)) >= 10
dds_star <- dds_star[keep,]

dds_star
```

#### Young vs Old

Now, we set the reference factor to young to get the differentially expressed genes between the young and the old group. As alpha we choose 0.05, like they did in the paper of Nativio et al.

```{r}
dds_star$group <- relevel(dds_star$group, ref = "Young")

dds_star_yo <- DESeq(dds_star)
res_star_yo <- results(dds_star_yo, alpha = 0.05)

res_star_yo
```

We take a look at the summary.

```{r}
summary(res_star_yo)
```

Finally, we generate a MA-plot to visualize the differentially expressed genes. With `lfcShrink` we use `type=ashr`, to separate the differentially and not differentially expressed genes more clearly.

```{r}
plotMA(res_star_yo, main="normal")

resAsh <- lfcShrink(dds_star_yo, coef=3, type="ashr")
plotMA(resAsh, main="ashr")
```

#### Old vs AD

We set the reference factor to old to get the differentially expressed genes between the old and the AD group this time.

```{r}
dds_star$group <- relevel(dds_star$group, ref = "Old")

dds_star_oad <- DESeq(dds_star)
res_star_oad <- results(dds_star_oad, alpha = 0.05)

res_star_oad
```

```{r}
summary(res_star_oad)
```

```{r}
plotMA(res_star_oad, main="normal")

resAsh <- lfcShrink(dds_star_oad, coef=3, type="ashr")
plotMA(resAsh, main="ashr")
```

#### AD vs Young

Lastly, we set the reference factor to old to get the differentially expressed genes between the AD and the young group.

```{r}
dds_star$group <- relevel(dds_star$group, ref = "AD")

dds_star_ady <- DESeq(dds_star)
res_star_ady <- results(dds_star_ady, alpha = 0.05)

res_star_ady
```

```{r}
summary(res_star_ady)
```

```{r}
plotMA(res_star_ady, main="normal")

resAsh <- lfcShrink(dds_star_ady, coef=3, type="ashr")
plotMA(resAsh, main="ashr")
```

#### General

In the next section, we compare between all three groups, so we apply each function only once.

In the first chunk, we take a look at the differences for the counts between the groups.

```{r}
plotCount <- plotCounts(
  dds_star_yo, 
  gene=which.min(res_star_yo$padj), 
  intgroup="group", 
  returnData=TRUE
  )

ggplot(plotCount, aes(x=group, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))
```

In the next step, a Variance Stabilizing Transformation was applied. `vst()` was chosen over `rld()` because it is a much faster transformation.

```{r}
msd <- meanSdPlot(assay(dds_star_yo), plot = FALSE)
msd$gg + 
  ggtitle('Without Transformation') 

vst <- vst(dds_star_yo, blind=FALSE)
msd_vst <- meanSdPlot(assay(vst), plot = FALSE)
msd_vst$gg + 
  ggtitle('Variance Stabilizing Transformation')
```

Now, we plot a heatmap for the Top 30 differentially expressed genes. Here we get an overview about the different expression levels between the groups.

```{r}
select <- order(rowMeans(
  counts(dds_star_yo,normalized=TRUE)),
  decreasing=TRUE
  )[1:30]

pheatmap(
  assay(vst)[select,],
  cluster_rows=TRUE,
  show_rownames=TRUE,
  cluster_cols=FALSE
  )
```

We build a sample distance matrix, to take a look at how the samples are clustered. This gives us an impression how similar the samples are within and in between the groups.

```{r}
sampleDists <- dist(t(assay(vst)))
sampleDistMatrix <- as.matrix(sampleDists)

rownames(sampleDistMatrix) <- paste(vst$group, sep="-")
colnames(sampleDistMatrix) <- NULL

colors <- colorRampPalette( rev(brewer.pal(9, "Reds")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

Lastly, we generate a PCA plot. This shows us again how reliable the samples are grouped or if there is a lot of intersection.

```{r}
plotPCA(vst, intgroup='group')
```

### kallisto

In this section, we will do the same steps for the count matrix where we aligned the data with kallisto. Again, we load our matrix. After taking a look at the head of the loaded data, it is clear that there needs to be done some tidying.

```{r}
kallisto <- read.table(
  'data/count.kallisto.txt', 
  header = TRUE, 
  sep = '\t',
  row.names = 1
  )

head(kallisto)
```

In order to get a clean matrix, we first drop the columns Chr, Start, End, Strand and Length as they are not necessary.

```{r}
kallisto <- kallisto[, -(1:5)]
```

The remaining columns correspond to the samples. However, they are still named after the path and filenames featureCounts received. To shorten the names and make the data easierto read, we define a list with the sample ids and then change the colnames into them. We now see that our data is much cleaner.

```{r}
colname <- list(
  '10-8A-Old','11-10T-Old','12-6A-Old','13-11T-Old','14-7A-Old','15-13T-Old',
  '16-14T-Old','17-9A-Old','18-10A-Old','19-11A-Old','20-1T-AD','21-1A-AD',
  '2-12A-Young','22-2T-AD','23-2A-AD','24-3T-AD','25-5T-AD','26-3A-AD',
  '27-5A-AD','28-8T-AD','29-6T-AD','30-9T-AD','31-7T-AD','3-17T-Young',
  '4-13A-Young','5-18T-Young','6-14A-Young','7-19T-Young','8-15A-Young','9-16A-Young'
  )

i <- 1
for (name in colnames(kallisto)){
  colnames(kallisto)[colnames(kallisto) == name] <- colname[i]
  i = i + 1
}

head(kallisto)
```

Now, we load the metafile we wrote for the kallisto data.

```{r}
coldata_kal <- read.csv(
  'data/kallisto_col.csv',
  row.names = 1,
  sep = ';'
  )

head(coldata_kal)
```

Again, we check the form of our data to make sure it is adequately prepared for DESeq.

```{r}
all(rownames(coldata_kal) %in% colnames(kallisto))
all(rownames(coldata_kal) == colnames(kallisto))
```

Now, we factor the groups again and build a DESeqDataSet afterwards.

```{r}
coldata_kal$group <- factor(coldata_kal$group)

dds_kallisto <- DESeqDataSetFromMatrix(
  countData = kallisto,
  colData = coldata_kal,
  design = ~ group
)

dds_kallisto
```

We pre-filter the DESeqDataSet. This time, about 36000 rows with low counts were removed. The kallisto DESeqDataSet had about 63000 counts in the beginning, while the STAR DESeqDataSet only had about 27000. It is assumed that Nativio et al. already did some pre-filtering on the STAR count matrix too.

```{r}
keep <- rowSums(counts(dds_kallisto)) >= 10
dds_kallisto <- dds_kallisto[keep,]

dds_kallisto
```

#### Young vs Old

We set the reference factor to young to get the differentially expressed genes between the young and the old group and perform everything accordingly to the STAR data.

```{r}
dds_kallisto$group <- relevel(dds_kallisto$group, ref = "Young")

dds_kal_yo <- DESeq(dds_kallisto)
res_kal_yo <- results(dds_kal_yo, alpha = 0.05)

res_kal_yo
```

```{r}
summary(res_kal_yo)
```

```{r}
plotMA(res_kal_yo, main="normal")

resAsh <- lfcShrink(dds_kal_yo, coef=3, type="ashr")
plotMA(resAsh, main="ashr")
```

#### Old vs AD

The same was done for the old and the AD group.

```{r}
dds_kallisto$group <- relevel(dds_kallisto$group, ref = "Old")

dds_kal_oad <- DESeq(dds_kallisto)
res_kal_oad <- results(dds_kal_oad, alpha = 0.05)

res_kal_oad
```

```{r}
summary(res_kal_oad)
```

```{r}
plotMA(res_kal_oad, main="normal")

resAsh <- lfcShrink(dds_kal_oad, coef=3, type="ashr")
plotMA(resAsh, main="ashr")
```

#### AD vs Young

The same was done for the AD and the young group.

```{r}
dds_kallisto$group <- relevel(dds_kallisto$group, ref = "AD")

dds_kal_ady <- DESeq(dds_kallisto)
res_kal_ady <- results(dds_kal_ady, alpha = 0.05)

res_kal_ady
```

```{r}
summary(res_kal_ady)
```

```{r}
plotMA(res_kal_ady, main="normal")

resAsh <- lfcShrink(dds_kal_ady, coef=3, type="ashr")
plotMA(resAsh, main="ashr")
```

#### General

Again, the general part was done only once as a comparison over all three groups.

```{r}
plotCount <- plotCounts(
  dds_kal_yo, 
  gene=which.min(res_kal_yo$padj), 
  intgroup="group", 
  returnData=TRUE
  )


ggplot(plotCount, aes(x=group, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))
```

```{r}
msd <- meanSdPlot(assay(dds_kal_yo), plot = FALSE)
msd$gg + 
  ggtitle('Without Transformation') 

vst <- vst(dds_kal_yo, blind=FALSE)
msd_vst <- meanSdPlot(assay(vst), plot = FALSE)
msd_vst$gg + 
  ggtitle('Variance Stabilizing Transformation')
```

For the heatmap, we needed to get the gene symbols from Ensembl, because we just had the gene ids in the kallisto count matrix.

```{r}
select <- order(rowMeans(
  counts(dds_kal_yo,normalized=TRUE)),
  decreasing=TRUE
  )[1:30]

library(EnsDb.Hsapiens.v79)
names <- ensembldb::select(
  EnsDb.Hsapiens.v79, 
  keys = rownames(assay(vst)[select,]), 
  keytype = "GENEID", 
  columns = c("SYMBOL","GENEID"))
print(names$SYMBOL)

pheatmap(
  assay(vst)[select,],
  cluster_rows = TRUE,
  show_rownames = TRUE,
  labels_row = names$SYMBOL,
  cluster_cols = FALSE
  )
```

```{r}
sampleDists <- dist(t(assay(vst)))
sampleDistMatrix <- as.matrix(sampleDists)

rownames(sampleDistMatrix) <- paste(vst$group, sep="-")
colnames(sampleDistMatrix) <- NULL

colors <- colorRampPalette( rev(brewer.pal(9, "Reds")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

```{r}
plotPCA(vst, intgroup='group')
```



## Machine Learning

The clustering part for the Machine Learning can be found in the following section. The generation of the ChIP-seq count matrices and the classification part can be found in the Python notebook [notebook.ipynb](https://github.com/nepsygirl/DataScience_finalProjekt_Group2/blob/main/notebook.ipynb) in the repository, in the section ***Machine Learning***.

## SVM method for RNA

This was the first try to implement the SVM for the RNA-seq data. Unfortunately, this did not work and was abadoned. The SVM was implemented with scikit-learn in the Python Notebook.

Because this did not work, the chunks in this section will not be evaluated.

```{r eval = FALSE}
#read rna count_matrix
rna_seq <- read.table(
  '/home/melika/analysis/GSE159699_summary_count.star.txt', 
  header = TRUE, 
  sep = '\t',
  row.names = 'refGene'
)
#head(rna_seq)

#read metafile
metafile<- read.csv(
  '/home/melika/analysis/star_col.csv',
  row.names = 1,
  sep = ';'
)
#head(metafile)
```

```{r eval = FALSE}
library(DESeq2)
library(RandPro)
library(caret)
library(rlang)
set.seed(2128)

i <- 1
for (name in colnames(rna_seq)){
  colnames(rna_seq)[colnames(rna_seq) == name] <- rownames(metafile)[i]
  i = i + 1
}

# check form of data
all(rownames(metafile) %in% colnames(rna_seq))
all(rownames(metafile) == colnames(rna_seq))

metafile$group <- factor(metafile$group)

#create DESeq matrix
de_matrix <- DESeqDataSetFromMatrix(
  countData = rna_seq,
  colData = metafile,
  design = ~ group
)

#if count<10 left out that row
keep <- rowSums(counts(de_matrix)) >= 10
de_matrix <- de_matrix[keep,]

#filtering genes based on Cook's distance when there are many degrees of freedom and create DESeq object 
de_matrix <- DESeq(de_matrix)

#sorts the values of matrix given by their index in increasing order
select <- order(rowMeans(
  counts(de_matrix, normalized=TRUE)),
  decreasing=TRUE
)[1:30]

# normalization of deseq_matrix
vst <- vst(de_matrix, blind=FALSE)
#extract normalized counts in readable format from the DEseq object 
data <- assay(vst)[select,]

#splitting data
nTest <- ceiling(ncol(data) * 0.7)
ind <- sample(ncol(data), nTest, FALSE)
```

```{r eval = FALSE}
#classification models.
train <- data[ ,ind] + 1
test <- data[ ,-ind] + 1
```

```{r eval = FALSE}
##did not work!!!

## Support vector machines with radial basis function kernel

set.seed(2128)

svm <- classify(train,test,method = "svmRadial",preProcessing = "deseq-vst")
show(svm)
trained(svm)
plot(svm)
```

### Hierachical Clustering

```{r}
#read rna count_matrix
rna <- read.table(
  '/home/melika/analysis/GSE159699_summary_count.star.txt', 
  header = TRUE, 
  sep = '\t',
  row.names = 'refGene' 
)
head(rna)

#read metafile
metafile <- read.csv(
  '/home/melika/analysis/star_col.csv',
  row.names = 1,
  sep = ';'
)
head(metafile)


i <- 1
for (name in colnames(rna)){
  colnames(rna)[colnames(rna) == name] <- rownames(metafile)[i]
  i = i + 1
}

# check form of data
all(rownames(metafile) %in% colnames(rna))
all(rownames(metafile) == colnames(rna))

metafile$group <- factor(metafile$group)
```

```{r}
library(DESeq2)
#create DESeq matrix
deseq_matrix <- DESeqDataSetFromMatrix(
  countData = rna,
  colData = metafile,
  design = ~ group
)

#filtering genes based on Cook's distance when there are many degrees of freedom and create DESeq object 
deseq_matrix <- DESeq(deseq_matrix)

# normalization of deseq_matrix
norm_matrix <- vst(dds_star)

#extract normalized counts in readable format from the DEseq object 
data <- assay(norm_matrix)
# save a table with the normalized data
write.table(data, sep="\t",file="Norm_data_rna.txt", row.names=TRUE,col.names=NA,quote=FALSE)
```

```{r}
#creat distance matrix 
sampleDists <- dist(t(data),  method = "euclidean")

#perform hierarchical clustering based on the distances matrix of RNA
library(rafalib)
rna_clu <- hclust(sampleDists)

#plot Dendrogram which shows the hierarchical clustering of RNA-seq
plot(rna_clu,labels=metafile$group,cex=0.5)

#add color to the cluters
myplclust(rna_clu, labels=metafile$group, lab.col=as.fumeric(as.vector(metafile$group)),cex=0.5)
```

```{r}
#cut the tree into clusters, 
#we can examine how the clusters overlap with our sample
hcluster <- cutree(rna_clu, h=30)
table(true=metafile$group, cluster=hcluster)
```

```{r}
#heatmap
library(RColorBrewer) 
library(genefilter)

hmcol <- colorRampPalette(brewer.pal(7,"YlOrBr"))(100)

rv <- rowVars(data)
#pick all the samples
idx <- order(-rv)[1:30]

library(gplots) ##Available from CRAN
head(cbind(colnames(data),cols))
cols <- palette(brewer.pal(8,"Dark2"))[as.fumeric(as.vector(metafile$group))]
heatmap.2(data[idx,], labCol=metafile$group,
          trace="none", 
          ColSideColors=cols, 
          col=hmcol)
```


## STRING Analysis

The STRING Analysis for the Top 30 genes of both RNA-seq datasets were done directly in the browser on the [STRING website](https://string-db.org/).