{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6678a3bf-e7b9-4c41-aaa3-9a71d026376c",
   "metadata": {},
   "source": [
    "# Data Science | Final Project | Group 02\n",
    "\n",
    "## Identify epigenetic alterations associated with Alzheimer’s disease and classification of gene expressions between healthy and sick patients\n",
    "\n",
    "#### Carmen Calle Huerta, Christina Kirschbaum, Pushpa Koirala, Melika Moradi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7b47320-9337-494f-a0af-932f496c4ca3",
   "metadata": {},
   "source": [
    "## Our Project\n",
    "\n",
    "Alzheimer’s disease is the most prevalent kind of dementia and a fatal brain ailment. More study into this illness might lead to a better understanding of the condition and more effective treatment options. \n",
    "\n",
    "In this project, ChIP-seq data for H3K27ac, H3K9ac, H3K122ac and H3K4me1 as well as RNA-seq data from the from the lateral temporal lobe of the human brain for young healthy patients, old heathy patients and patients with Alzheimers disease will be analyzed and the differences between normal aging and cognitive impairment will be explored. The epigenomic or transcriptomic profiles will be analyzed to find relevant epigenetic alterations associated with the disease and help to better understand the molecular pathophysiology underlying.\n",
    "\n",
    "Afterwards, with Machine Learning models the presence and absence of Alzheimer’s disease based on the data. The models will be built with Support Vector Machines and Random Forest.\n",
    "\n",
    "Finally, the findings will be compared to them in related papers, we will look into relevant *in vivo* experiments with model organisms and use the Genome Browser to generate ChIP-seq tracks.\n",
    "\n",
    "For our project we were inspirated by the paper:\n",
    "Nativio R, Lan Y, Donahue G et al. [\"An integrated multi-omics approach identifies\n",
    "epigenetic alterations associated with Alzheimer’s disease.\"](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8098004/)\n",
    "\n",
    "The data is derived from GEO, [Series GSE153875](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE153875)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e7a13d3",
   "metadata": {},
   "source": [
    "## Import packages\n",
    "\n",
    "First, we import some of the most frequently used packages in Python. [NumPy](https://numpy.org/doc/stable/) for working with arrays, matrices and linear algebra, [pandas](https://pandas.pydata.org/docs/) for data analysis and manipulation, [matplotlib](https://matplotlib.org/) and [seaborn](https://seaborn.pydata.org/) for visualizations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d313e8ca",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a67bd37d-c0e3-41d2-98ec-f7d302110cf1",
   "metadata": {},
   "source": [
    "## Data Integration and Preprocessing RNA-seq\n",
    "\n",
    "In the next step, we import our RNA-seq data from the SubSeries GSE159699 of the SuperSeries GSE153875 with SRA-Toolkit and fastq-dump. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d844c8bf-c638-4e31-85c2-6bb4b16ba761",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently downloading: SRR12850830\n",
      "The command used was: prefetch SRR12850830\n",
      "\n",
      "2022-06-13T18:59:40 prefetch.2.8.2: 1) Downloading 'SRR12850830'...\n",
      "2022-06-13T18:59:40 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:02:10 prefetch.2.8.2: 1) 'SRR12850830' was downloaded successfully\n",
      "2022-06-13T19:02:10 prefetch.2.8.2: 'SRR12850830' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850831\n",
      "The command used was: prefetch SRR12850831\n",
      "\n",
      "2022-06-13T19:02:12 prefetch.2.8.2: 1) Downloading 'SRR12850831'...\n",
      "2022-06-13T19:02:12 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:04:46 prefetch.2.8.2: 1) 'SRR12850831' was downloaded successfully\n",
      "2022-06-13T19:04:46 prefetch.2.8.2: 'SRR12850831' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850832\n",
      "The command used was: prefetch SRR12850832\n",
      "\n",
      "2022-06-13T19:04:48 prefetch.2.8.2: 1) Downloading 'SRR12850832'...\n",
      "2022-06-13T19:04:48 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:07:13 prefetch.2.8.2: 1) 'SRR12850832' was downloaded successfully\n",
      "2022-06-13T19:07:13 prefetch.2.8.2: 'SRR12850832' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850833\n",
      "The command used was: prefetch SRR12850833\n",
      "\n",
      "2022-06-13T19:07:15 prefetch.2.8.2: 1) Downloading 'SRR12850833'...\n",
      "2022-06-13T19:07:15 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:10:02 prefetch.2.8.2: 1) 'SRR12850833' was downloaded successfully\n",
      "2022-06-13T19:10:02 prefetch.2.8.2: 'SRR12850833' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850834\n",
      "The command used was: prefetch SRR12850834\n",
      "\n",
      "2022-06-13T19:10:05 prefetch.2.8.2: 1) Downloading 'SRR12850834'...\n",
      "2022-06-13T19:10:05 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:12:18 prefetch.2.8.2: 1) 'SRR12850834' was downloaded successfully\n",
      "2022-06-13T19:12:18 prefetch.2.8.2: 'SRR12850834' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850835\n",
      "The command used was: prefetch SRR12850835\n",
      "\n",
      "2022-06-13T19:12:20 prefetch.2.8.2: 1) Downloading 'SRR12850835'...\n",
      "2022-06-13T19:12:20 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:14:38 prefetch.2.8.2: 1) 'SRR12850835' was downloaded successfully\n",
      "2022-06-13T19:14:38 prefetch.2.8.2: 'SRR12850835' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850836\n",
      "The command used was: prefetch SRR12850836\n",
      "\n",
      "2022-06-13T19:14:41 prefetch.2.8.2: 1) Downloading 'SRR12850836'...\n",
      "2022-06-13T19:14:41 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:17:07 prefetch.2.8.2: 1) 'SRR12850836' was downloaded successfully\n",
      "2022-06-13T19:17:07 prefetch.2.8.2: 'SRR12850836' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850837\n",
      "The command used was: prefetch SRR12850837\n",
      "\n",
      "2022-06-13T19:17:09 prefetch.2.8.2: 1) Downloading 'SRR12850837'...\n",
      "2022-06-13T19:17:09 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:19:37 prefetch.2.8.2: 1) 'SRR12850837' was downloaded successfully\n",
      "2022-06-13T19:19:37 prefetch.2.8.2: 'SRR12850837' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850838\n",
      "The command used was: prefetch SRR12850838\n",
      "\n",
      "2022-06-13T19:19:40 prefetch.2.8.2: 1) Downloading 'SRR12850838'...\n",
      "2022-06-13T19:19:40 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:21:58 prefetch.2.8.2: 1) 'SRR12850838' was downloaded successfully\n",
      "2022-06-13T19:21:58 prefetch.2.8.2: 'SRR12850838' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850839\n",
      "The command used was: prefetch SRR12850839\n",
      "\n",
      "2022-06-13T19:22:01 prefetch.2.8.2: 1) Downloading 'SRR12850839'...\n",
      "2022-06-13T19:22:01 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:24:45 prefetch.2.8.2: 1) 'SRR12850839' was downloaded successfully\n",
      "2022-06-13T19:24:45 prefetch.2.8.2: 'SRR12850839' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850840\n",
      "The command used was: prefetch SRR12850840\n",
      "\n",
      "2022-06-13T19:24:48 prefetch.2.8.2: 1) Downloading 'SRR12850840'...\n",
      "2022-06-13T19:24:48 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:27:12 prefetch.2.8.2: 1) 'SRR12850840' was downloaded successfully\n",
      "2022-06-13T19:27:12 prefetch.2.8.2: 'SRR12850840' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850841\n",
      "The command used was: prefetch SRR12850841\n",
      "\n",
      "2022-06-13T19:27:14 prefetch.2.8.2: 1) Downloading 'SRR12850841'...\n",
      "2022-06-13T19:27:14 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:29:29 prefetch.2.8.2: 1) 'SRR12850841' was downloaded successfully\n",
      "2022-06-13T19:29:29 prefetch.2.8.2: 'SRR12850841' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850842\n",
      "The command used was: prefetch SRR12850842\n",
      "\n",
      "2022-06-13T19:29:32 prefetch.2.8.2: 1) Downloading 'SRR12850842'...\n",
      "2022-06-13T19:29:32 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:32:05 prefetch.2.8.2: 1) 'SRR12850842' was downloaded successfully\n",
      "2022-06-13T19:32:05 prefetch.2.8.2: 'SRR12850842' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850843\n",
      "The command used was: prefetch SRR12850843\n",
      "\n",
      "2022-06-13T19:32:08 prefetch.2.8.2: 1) Downloading 'SRR12850843'...\n",
      "2022-06-13T19:32:08 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:34:35 prefetch.2.8.2: 1) 'SRR12850843' was downloaded successfully\n",
      "2022-06-13T19:34:36 prefetch.2.8.2: 'SRR12850843' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850844\n",
      "The command used was: prefetch SRR12850844\n",
      "\n",
      "2022-06-13T19:34:38 prefetch.2.8.2: 1) Downloading 'SRR12850844'...\n",
      "2022-06-13T19:34:38 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:37:12 prefetch.2.8.2: 1) 'SRR12850844' was downloaded successfully\n",
      "2022-06-13T19:37:12 prefetch.2.8.2: 'SRR12850844' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850845\n",
      "The command used was: prefetch SRR12850845\n",
      "\n",
      "2022-06-13T19:37:14 prefetch.2.8.2: 1) Downloading 'SRR12850845'...\n",
      "2022-06-13T19:37:14 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:40:27 prefetch.2.8.2: 1) 'SRR12850845' was downloaded successfully\n",
      "2022-06-13T19:40:27 prefetch.2.8.2: 'SRR12850845' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850846\n",
      "The command used was: prefetch SRR12850846\n",
      "\n",
      "2022-06-13T19:40:29 prefetch.2.8.2: 1) Downloading 'SRR12850846'...\n",
      "2022-06-13T19:40:29 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:42:23 prefetch.2.8.2: 1) 'SRR12850846' was downloaded successfully\n",
      "2022-06-13T19:42:23 prefetch.2.8.2: 'SRR12850846' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850847\n",
      "The command used was: prefetch SRR12850847\n",
      "\n",
      "2022-06-13T19:42:26 prefetch.2.8.2: 1) Downloading 'SRR12850847'...\n",
      "2022-06-13T19:42:26 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:44:53 prefetch.2.8.2: 1) 'SRR12850847' was downloaded successfully\n",
      "2022-06-13T19:44:53 prefetch.2.8.2: 'SRR12850847' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850848\n",
      "The command used was: prefetch SRR12850848\n",
      "\n",
      "2022-06-13T19:44:56 prefetch.2.8.2: 1) Downloading 'SRR12850848'...\n",
      "2022-06-13T19:44:56 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:47:16 prefetch.2.8.2: 1) 'SRR12850848' was downloaded successfully\n",
      "2022-06-13T19:47:16 prefetch.2.8.2: 'SRR12850848' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850849\n",
      "The command used was: prefetch SRR12850849\n",
      "\n",
      "2022-06-13T19:47:19 prefetch.2.8.2: 1) Downloading 'SRR12850849'...\n",
      "2022-06-13T19:47:19 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:49:07 prefetch.2.8.2: 1) 'SRR12850849' was downloaded successfully\n",
      "2022-06-13T19:49:07 prefetch.2.8.2: 'SRR12850849' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850850\n",
      "The command used was: prefetch SRR12850850\n",
      "\n",
      "2022-06-13T19:49:09 prefetch.2.8.2: 1) Downloading 'SRR12850850'...\n",
      "2022-06-13T19:49:09 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:51:19 prefetch.2.8.2: 1) 'SRR12850850' was downloaded successfully\n",
      "2022-06-13T19:51:19 prefetch.2.8.2: 'SRR12850850' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850851\n",
      "The command used was: prefetch SRR12850851\n",
      "\n",
      "2022-06-13T19:51:21 prefetch.2.8.2: 1) Downloading 'SRR12850851'...\n",
      "2022-06-13T19:51:21 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:53:51 prefetch.2.8.2: 1) 'SRR12850851' was downloaded successfully\n",
      "2022-06-13T19:53:51 prefetch.2.8.2: 'SRR12850851' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850852\n",
      "The command used was: prefetch SRR12850852\n",
      "\n",
      "2022-06-13T19:53:54 prefetch.2.8.2: 1) Downloading 'SRR12850852'...\n",
      "2022-06-13T19:53:54 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:56:24 prefetch.2.8.2: 1) 'SRR12850852' was downloaded successfully\n",
      "2022-06-13T19:56:24 prefetch.2.8.2: 'SRR12850852' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850853\n",
      "The command used was: prefetch SRR12850853\n",
      "\n",
      "2022-06-13T19:56:27 prefetch.2.8.2: 1) Downloading 'SRR12850853'...\n",
      "2022-06-13T19:56:27 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T19:58:06 prefetch.2.8.2: 1) 'SRR12850853' was downloaded successfully\n",
      "2022-06-13T19:58:06 prefetch.2.8.2: 'SRR12850853' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850854\n",
      "The command used was: prefetch SRR12850854\n",
      "\n",
      "2022-06-13T19:58:09 prefetch.2.8.2: 1) Downloading 'SRR12850854'...\n",
      "2022-06-13T19:58:09 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T20:00:20 prefetch.2.8.2: 1) 'SRR12850854' was downloaded successfully\n",
      "2022-06-13T20:00:20 prefetch.2.8.2: 'SRR12850854' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850855\n",
      "The command used was: prefetch SRR12850855\n",
      "\n",
      "2022-06-13T20:00:22 prefetch.2.8.2: 1) Downloading 'SRR12850855'...\n",
      "2022-06-13T20:00:22 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T20:02:38 prefetch.2.8.2: 1) 'SRR12850855' was downloaded successfully\n",
      "2022-06-13T20:02:38 prefetch.2.8.2: 'SRR12850855' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850856\n",
      "The command used was: prefetch SRR12850856\n",
      "\n",
      "2022-06-13T20:02:40 prefetch.2.8.2: 1) Downloading 'SRR12850856'...\n",
      "2022-06-13T20:02:40 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T20:04:48 prefetch.2.8.2: 1) 'SRR12850856' was downloaded successfully\n",
      "2022-06-13T20:04:48 prefetch.2.8.2: 'SRR12850856' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850857\n",
      "The command used was: prefetch SRR12850857\n",
      "\n",
      "2022-06-13T20:04:51 prefetch.2.8.2: 1) Downloading 'SRR12850857'...\n",
      "2022-06-13T20:04:51 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T20:07:18 prefetch.2.8.2: 1) 'SRR12850857' was downloaded successfully\n",
      "2022-06-13T20:07:18 prefetch.2.8.2: 'SRR12850857' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850858\n",
      "The command used was: prefetch SRR12850858\n",
      "\n",
      "2022-06-13T20:07:21 prefetch.2.8.2: 1) Downloading 'SRR12850858'...\n",
      "2022-06-13T20:07:21 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T20:09:22 prefetch.2.8.2: 1) 'SRR12850858' was downloaded successfully\n",
      "2022-06-13T20:09:22 prefetch.2.8.2: 'SRR12850858' has 0 unresolved dependencies\n",
      "Currently downloading: SRR12850859\n",
      "The command used was: prefetch SRR12850859\n",
      "\n",
      "2022-06-13T20:09:24 prefetch.2.8.2: 1) Downloading 'SRR12850859'...\n",
      "2022-06-13T20:09:24 prefetch.2.8.2:  Downloading via https...\n",
      "2022-06-13T20:11:49 prefetch.2.8.2: 1) 'SRR12850859' was downloaded successfully\n",
      "2022-06-13T20:11:49 prefetch.2.8.2: 'SRR12850859' has 0 unresolved dependencies\n",
      "Generating fastq for: SRR12850830\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850830.sra\n",
      "Rejected 37552625 READS because of filtering out non-biological READS\n",
      "Read 37552625 spots for /home/christina/ncbi/public/sra/SRR12850830.sra\n",
      "Written 37552625 spots for /home/christina/ncbi/public/sra/SRR12850830.sra\n",
      "Generating fastq for: SRR12850831\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850831.sra\n",
      "Rejected 38052606 READS because of filtering out non-biological READS\n",
      "Generating fastq for: SRR12850832\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850832.sra\n",
      "Read 38052606 spots for /home/christina/ncbi/public/sra/SRR12850831.sra\n",
      "Written 38052606 spots for /home/christina/ncbi/public/sra/SRR12850831.sra\n",
      "Rejected 36212177 READS because of filtering out non-biological READS\n",
      "Read 36212177 spots for /home/christina/ncbi/public/sra/SRR12850832.sra\n",
      "Written 36212177 spots for /home/christina/ncbi/public/sra/SRR12850832.sra\n",
      "Generating fastq for: SRR12850833\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850833.sra\n",
      "Rejected 42174836 READS because of filtering out non-biological READS\n",
      "Read 42174836 spots for /home/christina/ncbi/public/sra/SRR12850833.sra\n",
      "Written 42174836 spots for /home/christina/ncbi/public/sra/SRR12850833.sra\n",
      "Generating fastq for: SRR12850834\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850834.sra\n",
      "Rejected 38610276 READS because of filtering out non-biological READS\n",
      "Read 38610276 spots for /home/christina/ncbi/public/sra/SRR12850834.sra\n",
      "Written 38610276 spots for /home/christina/ncbi/public/sra/SRR12850834.sra\n",
      "Generating fastq for: SRR12850835\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850835.sra\n",
      "Rejected 40295518 READS because of filtering out non-biological READS\n",
      "Read 40295518 spots for /home/christina/ncbi/public/sra/SRR12850835.sra\n",
      "Written 40295518 spots for /home/christina/ncbi/public/sra/SRR12850835.sra\n",
      "Generating fastq for: SRR12850836\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850836.sra\n",
      "Rejected 40036785 READS because of filtering out non-biological READS\n",
      "Read 40036785 spots for /home/christina/ncbi/public/sra/SRR12850836.sra\n",
      "Written 40036785 spots for /home/christina/ncbi/public/sra/SRR12850836.sra\n",
      "Generating fastq for: SRR12850837\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850837.sra\n",
      "Rejected 39919539 READS because of filtering out non-biological READS\n",
      "Read 39919539 spots for /home/christina/ncbi/public/sra/SRR12850837.sra\n",
      "Written 39919539 spots for /home/christina/ncbi/public/sra/SRR12850837.sra\n",
      "Generating fastq for: SRR12850838\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850838.sra\n",
      "Rejected 36990438 READS because of filtering out non-biological READS\n",
      "Read 36990438 spots for /home/christina/ncbi/public/sra/SRR12850838.sra\n",
      "Written 36990438 spots for /home/christina/ncbi/public/sra/SRR12850838.sra\n",
      "Generating fastq for: SRR12850839\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850839.sra\n",
      "Rejected 43899259 READS because of filtering out non-biological READS\n",
      "Read 43899259 spots for /home/christina/ncbi/public/sra/SRR12850839.sra\n",
      "Written 43899259 spots for /home/christina/ncbi/public/sra/SRR12850839.sra\n",
      "Generating fastq for: SRR12850840\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850840.sra\n",
      "Rejected 40299174 READS because of filtering out non-biological READS\n",
      "Read 40299174 spots for /home/christina/ncbi/public/sra/SRR12850840.sra\n",
      "Written 40299174 spots for /home/christina/ncbi/public/sra/SRR12850840.sra\n",
      "Generating fastq for: SRR12850841\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850841.sra\n",
      "Rejected 39739040 READS because of filtering out non-biological READS\n",
      "Read 39739040 spots for /home/christina/ncbi/public/sra/SRR12850841.sra\n",
      "Written 39739040 spots for /home/christina/ncbi/public/sra/SRR12850841.sra\n",
      "Generating fastq for: SRR12850842\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850842.sra\n",
      "Rejected 39471683 READS because of filtering out non-biological READS\n",
      "Read 39471683 spots for /home/christina/ncbi/public/sra/SRR12850842.sra\n",
      "Written 39471683 spots for /home/christina/ncbi/public/sra/SRR12850842.sra\n",
      "Generating fastq for: SRR12850843\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850843.sra\n",
      "Rejected 39459711 READS because of filtering out non-biological READS\n",
      "Read 39459711 spots for /home/christina/ncbi/public/sra/SRR12850843.sra\n",
      "Written 39459711 spots for /home/christina/ncbi/public/sra/SRR12850843.sra\n",
      "Generating fastq for: SRR12850844\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850844.sra\n",
      "Rejected 38019285 READS because of filtering out non-biological READS\n",
      "Read 38019285 spots for /home/christina/ncbi/public/sra/SRR12850844.sra\n",
      "Written 38019285 spots for /home/christina/ncbi/public/sra/SRR12850844.sra\n",
      "Generating fastq for: SRR12850845\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850845.sra\n",
      "Rejected 40847873 READS because of filtering out non-biological READS\n",
      "Read 40847873 spots for /home/christina/ncbi/public/sra/SRR12850845.sra\n",
      "Written 40847873 spots for /home/christina/ncbi/public/sra/SRR12850845.sra\n",
      "Generating fastq for: SRR12850846\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850846.sra\n",
      "Rejected 35468886 READS because of filtering out non-biological READS\n",
      "Read 35468886 spots for /home/christina/ncbi/public/sra/SRR12850846.sra\n",
      "Written 35468886 spots for /home/christina/ncbi/public/sra/SRR12850846.sra\n",
      "Generating fastq for: SRR12850847\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850847.sra\n",
      "Rejected 36432788 READS because of filtering out non-biological READS\n",
      "Read 36432788 spots for /home/christina/ncbi/public/sra/SRR12850847.sra\n",
      "Written 36432788 spots for /home/christina/ncbi/public/sra/SRR12850847.sra\n",
      "Generating fastq for: SRR12850848\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850848.sra\n",
      "Rejected 38367345 READS because of filtering out non-biological READS\n",
      "Read 38367345 spots for /home/christina/ncbi/public/sra/SRR12850848.sra\n",
      "Written 38367345 spots for /home/christina/ncbi/public/sra/SRR12850848.sra\n",
      "Generating fastq for: SRR12850849\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850849.sra\n",
      "Rejected 31893808 READS because of filtering out non-biological READS\n",
      "Read 31893808 spots for /home/christina/ncbi/public/sra/SRR12850849.sra\n",
      "Written 31893808 spots for /home/christina/ncbi/public/sra/SRR12850849.sra\n",
      "Generating fastq for: SRR12850850\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850850.sra\n",
      "Rejected 36151808 READS because of filtering out non-biological READS\n",
      "Read 36151808 spots for /home/christina/ncbi/public/sra/SRR12850850.sra\n",
      "Written 36151808 spots for /home/christina/ncbi/public/sra/SRR12850850.sra\n",
      "Generating fastq for: SRR12850851\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850851.sra\n",
      "Rejected 40063230 READS because of filtering out non-biological READS\n",
      "Read 40063230 spots for /home/christina/ncbi/public/sra/SRR12850851.sra\n",
      "Written 40063230 spots for /home/christina/ncbi/public/sra/SRR12850851.sra\n",
      "Generating fastq for: SRR12850852\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850852.sra\n",
      "Rejected 39847390 READS because of filtering out non-biological READS\n",
      "Read 39847390 spots for /home/christina/ncbi/public/sra/SRR12850852.sra\n",
      "Written 39847390 spots for /home/christina/ncbi/public/sra/SRR12850852.sra\n",
      "Generating fastq for: SRR12850853\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850853.sra\n",
      "Rejected 26043105 READS because of filtering out non-biological READS\n",
      "Read 26043105 spots for /home/christina/ncbi/public/sra/SRR12850853.sra\n",
      "Written 26043105 spots for /home/christina/ncbi/public/sra/SRR12850853.sra\n",
      "Generating fastq for: SRR12850854\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850854.sra\n",
      "Rejected 35768170 READS because of filtering out non-biological READS\n",
      "Read 35768170 spots for /home/christina/ncbi/public/sra/SRR12850854.sra\n",
      "Written 35768170 spots for /home/christina/ncbi/public/sra/SRR12850854.sra\n",
      "Generating fastq for: SRR12850855\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850855.sra\n",
      "Rejected 38207241 READS because of filtering out non-biological READS\n",
      "Read 38207241 spots for /home/christina/ncbi/public/sra/SRR12850855.sra\n",
      "Written 38207241 spots for /home/christina/ncbi/public/sra/SRR12850855.sra\n",
      "Generating fastq for: SRR12850856\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850856.sra\n",
      "Rejected 38638267 READS because of filtering out non-biological READS\n",
      "Read 38638267 spots for /home/christina/ncbi/public/sra/SRR12850856.sra\n",
      "Written 38638267 spots for /home/christina/ncbi/public/sra/SRR12850856.sra\n",
      "Generating fastq for: SRR12850857\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850857.sra\n",
      "Rejected 38465281 READS because of filtering out non-biological READS\n",
      "Read 38465281 spots for /home/christina/ncbi/public/sra/SRR12850857.sra\n",
      "Written 38465281 spots for /home/christina/ncbi/public/sra/SRR12850857.sra\n",
      "Generating fastq for: SRR12850858\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850858.sra\n",
      "Rejected 30568266 READS because of filtering out non-biological READS\n",
      "Read 30568266 spots for /home/christina/ncbi/public/sra/SRR12850858.sra\n",
      "Written 30568266 spots for /home/christina/ncbi/public/sra/SRR12850858.sra\n",
      "Generating fastq for: SRR12850859\n",
      "The command used was: fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/SRR12850859.sra\n",
      "Rejected 36817470 READS because of filtering out non-biological READS\n",
      "Read 36817470 spots for /home/christina/ncbi/public/sra/SRR12850859.sra\n",
      "Written 36817470 spots for /home/christina/ncbi/public/sra/SRR12850859.sra\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# sra numbers of our RNA-seq data from SuperSeries GSE153875 / SubSeries GSE159699\n",
    "sra_numbers = [\n",
    "    \"SRR12850830\", \"SRR12850831\", \"SRR12850832\", \"SRR12850833\", \"SRR12850834\",\n",
    "    \"SRR12850835\", \"SRR12850836\", \"SRR12850837\", \"SRR12850838\", \"SRR12850839\",\n",
    "    \"SRR12850840\", \"SRR12850841\", \"SRR12850842\", \"SRR12850843\", \"SRR12850844\",\n",
    "    \"SRR12850845\", \"SRR12850846\", \"SRR12850847\", \"SRR12850848\", \"SRR12850849\",\n",
    "    \"SRR12850850\", \"SRR12850851\", \"SRR12850852\", \"SRR12850853\", \"SRR12850854\",\n",
    "    \"SRR12850855\", \"SRR12850856\", \"SRR12850857\", \"SRR12850858\", \"SRR12850859\"\n",
    "    ]\n",
    "\n",
    "# this will download the .sra files to ~/ncbi/public/sra/ \n",
    "for sra_id in sra_numbers:\n",
    "    print (\"Currently downloading: \" + sra_id)\n",
    "    prefetch = \"prefetch \" + sra_id\n",
    "    print (\"The command used was: \" + prefetch)\n",
    "    subprocess.call(prefetch, shell=True)\n",
    "\n",
    "# this will extract the .sra files from above into a folder named 'fastq'\n",
    "for sra_id in sra_numbers:\n",
    "    print (\"Generating fastq for: \" + sra_id)\n",
    "    fastq_dump = \"fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip ~/ncbi/public/sra/\" + sra_id + \".sra\"\n",
    "    print (\"The command used was: \" + fastq_dump)\n",
    "    subprocess.call(fastq_dump, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4f5fdba-0d9e-4a51-9c1e-0d7f9bbdd596",
   "metadata": {},
   "source": [
    "After we downloaded the data, we will now preprocess it with STAR, an aligner for RNA-seq data mapping. Like in the paper of *Nativio et al.*, we will align our RNA-seq reads to the human reference genome (assembly GRCh37.75/hg19) using STAR with default parameters.\n",
    "\n",
    "This will create .sam files out of the .fastq files.\n",
    "\n",
    "**Currently in work or running**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78a150c2-3234-4dad-a697-6288ee49256a",
   "metadata": {},
   "source": [
    "## Data Integration for ChIP-seq\n",
    "\n",
    "The .bed and .bw files for the ChIP-seq data were downloaded. They are available as supplementary files of the SuperSeries GSE153875 from GEO. We split them into folders for H3K27ac, H3K9ac, H3K122ac and H3K4me1 (and peaks) to get smaller sets.\n",
    "\n",
    "**Where to store the data?*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e548f025-56f7-4b10-8776-fa3f6f29742a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The following code can get the metadata (if needed) for the ChIP-seq data with the library GEOparse (similar to GEOquery).\n",
    "import GEOparse\n",
    "\n",
    "gse = GEOparse.get_GEO(geo=\"GSE153875\", destdir=\"./\")\n",
    "\n",
    "# prints the metadata for the first sample\n",
    "for gsm_name, gsm in gse.gsms.items():\n",
    "    print(\"Name: \", gsm_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gsm.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
